src/wp-admin/includes/upgrade.php

```diff
@@ -2248,6 +2252,64 @@ function upgrade_560() {
 	}
 }
 
+/**
+ * Executes changes made in WordPress 5.8.0.
+ *
+ * @ignore
+ * @since 5.8.0
+ */
+function upgrade_580() {
+	_upgrade_580_force_deactivate_incompatible_plugins();
+	if ( false === get_option( 'wp_force_deactivated_plugins' ) ) {
+		add_option( 'wp_force_deactivated_plugins', array() );
+	}
+	if ( false === get_site_option( 'wp_force_deactivated_plugins' ) ) {
+		add_site_option( 'wp_force_deactivated_plugins', array() );
+	}
+}
+
+/**
+ * Duplicate of function in `upgrade-core.php` as it's unavailable during upgrades.
+ *
+ * @since 5.8.0
+ * @ignore
+ */
+function _upgrade_580_force_deactivate_incompatible_plugins() {
+	if ( ! file_exists( WP_PLUGIN_DIR . '/gutenberg/gutenberg.php' ) ) {
+		// Gutenberg is not installed.
+		return;
+	}
+
+	$gutenberg_data    = get_plugin_data( WP_PLUGIN_DIR . '/gutenberg/gutenberg.php' );
+	$gutenberg_version = ! empty( $gutenberg_data['Version'] ) ? $gutenberg_data['Version'] : false;
+
+	if ( ! $gutenberg_version ) {
+		// This may not be Gutenberg.
+		return;
+	}
+
+	if ( ! version_compare( $gutenberg_version, '10.7', '<=' ) ) {
+		// This version of Gutenberg is compatible with WP 5.8.
+		return;
+	}
+
+	$deactivated_gutenberg['gutenberg'] = array(
+		'plugin_name'         => 'Gutenberg',
+		'version_deactivated' => $gutenberg_version,
+		'version_compatible'  => '10.8',
+	);
+	if ( is_plugin_active_for_network( 'gutenberg/gutenberg.php' ) ) {
```

Comment:
Something about this blog feels weird to me. I'm wondering if it's possible for a site to have activated a plugin and then for the plugin to have been activated network-wide and if so, what would happen here.  This might be a nonissue ( that network activating a plugin causes it to be deactivated on individual sites) or this might be such an edge case that we don't need to worry about. 


I think it might make sense to abstract this out a bit into something like a `wp_set_deactivated_plugin` function